Halos pipeline

trivialTZ opened the pull request:
@sibirrer 
This pull request primarily adds the functionality to generate halos. This includes `halos.py` (which is used to calculate the mass of halos generated at a corresponding redshift, and the number of halos to be generated), `get_halos.ipynb` (which demonstrates the halo generation pipeline), `halo.yml` and the corresponding `readme.md` file (mainly describing the parameters for halo generation), and `halos_pipeline.py` (which is used to read the .yml file). There are also some pytest test functions included.

It is important to note that running these files requires using a specific branch: `skypy/halos` (https://github.com/skypyproject/skypy/tree/module/halos) !!!!! (I believe this is the reason test failed)

trivialTZ committed:
pipeline for halos

Some bugs still need to be fixed

trivialTZ committed:
Add Halos pipeline

with documentation, jupyter sample, test function.

trivialTZ committed:
Create .name

sibirrer commented on sim_pipeline/halos/halos.py:
```diff
@@ -0,0 +1,235 @@
+
+    :param z: The redshift at which to evaluate the number density.
+    :type z: float
+    :param m_min: The minimum halo mass.
```
can you provide units for these quantities, e.g. M_sol or M_sol/h?

sibirrer commented on sim_pipeline/halos/halos.py:
```diff
@@ -0,0 +1,235 @@
+    :param z: The redshift at which to evaluate the growth factor.
+    :type z: float
+    :param cosmology: The cosmology instance to use.
+    :type cosmology: Cosmology, optional
```
I guess this is an ~astropy.cosmology instance?

sibirrer commented on sim_pipeline/halos/halos.py:
```diff
@@ -0,0 +1,235 @@
+    :rtype: array
+    """
+    # TODO: debug
+    if m_min is None:
```
these default values settings (I think) you have in two routines. Perhaps you can make one 'utility' function that provides these settings and can be called for two or more functions, to avoid duplication

trivialTZ replied:
@sibirrer See the new update for halos.py

sibirrer reviewed (changes requested):
This looks excellent. I have some minor requests.
The tests unfortunately fail because the skpypy halo module is not in the pip or main branch.
I try to push on the skpypy folks to push it to main, also linking to this PR

trivialTZ committed:
Update halos.py

Create a utility function that handles default parameter and then modify the other functions to use this new utility function.

sibirrer reviewed (commented):
Thanks a lot @trivialTZ ! I wait to hear back from the SkyPy team about the halo module on their side, such that we can hopefully execute the tests as well as check for the test coverage. Other than that it looks great!

trivialTZ committed:
Add Class HalosLens() foe computering lensing effect for halos

1. update Jupyter notebook with the computing gamma and kappa
2. change yml file to make it run quicker for debug(will change back)
3. add halos_len.py with Class HalosLens() in it and def concentration_from_mass() in it.

trivialTZ commented:
@sibirrer Add Class HalosLens() for computing lensing effect for halos:
1. update Jupyter notebook with the computing gamma and kappa.
2. change yml file to make it run quicker for debug(will change back).
3. add halos_len.py with Class HalosLens() in it and def concentration_from_mass() in it.

sibirrer commented on sim_pipeline/halos/get_halos.ipynb:
```diff
@@ -0,0 +1,166 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #5.    halos_lens = HalosLens(halos_list=halos, sky_area=0.00001)
</pre><p>might be worth making a cosmology instance such that it is well reproducible and no UserWarning is raised in the output line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1279307835'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

sibirrer commented on sim_pipeline/halos/get_halos.ipynb:
```diff
@@ -0,0 +1,166 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #8.    print('kappa at center is: ',kappa)
</pre><p>Great notebook! The kappa value seems very high. Did you subtract the mean kappa of random drawn distributions?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1279307843'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

trivialTZ replied:
Currently, I haven't performed the random draws, but I'll make sure to include them in my upcoming work.

trivialTZ replied:
> Line #8.    print('kappa at center is: ',kappa)
> Great notebook! The kappa value seems very high. Did you subtract the mean kappa of random drawn distributions?
> 
> 
> _Reply via [ReviewNB](https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1279307843)_

Currently, I haven't performed the random draws, but I'll make sure to include them in my upcoming work.

sibirrer commented on sim_pipeline/halos/halos_lens.py:
```diff
@@ -0,0 +1,196 @@
+            from astropy.cosmology import default_cosmology
+            self.cosmo = default_cosmology.get()
+
+    def set_lens_model(self):
```
does this class always operates with the same lens model or is it intended to create a new class instance each time you use it? Because how it is currently used, you generate every time a new class instance (which might slow things down a bit)

sibirrer commented on sim_pipeline/halos/halos_lens.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        phi = 2 * np.pi * np.random.random()
+        upper_bound = np.sqrt(self.sky_area / np.pi)
+        random_radius = 3600 * np.random.uniform(0, upper_bound)
```
does this render a uniform distribution within a disk? Uniform in radius is not uniform in area

trivialTZ replied:
uniform distribution for radius and phi for angle, combined for uniform in area

trivialTZ replied:
> does this render a uniform distribution within a disk? Uniform in radius is not uniform in area

uniform distribution for radius and phi for angle, combined for uniform in area

sibirrer replied:
I am confused, uniform in radius will lead to non-uniform in area. See e.g. here how you can render uniform in area within a radius: https://github.com/LSST-strong-lensing/sim-pipeline/blob/1ef82e2094e5a9e4a6f42fcdf66acc0aae84b2d7/sim_pipeline/gg_lens.py#L78

sibirrer replied:
this might also be part of the problem as you sample too many halos too close to the center

trivialTZ replied:
I see, I am too presumptuous for it. That's right!

sibirrer reviewed (changes requested):
This is excellent coding and implementation. I have a few comments and hope that skypy merges the halo branch soon

trivialTZ committed:
Major update for halos+Glass weak lensing computing

1. **halos_lens.py**:
   - Introduced `get_kappa_gamma_distib()` for computing the kappa and gamma_tot values from the list.
   - Implemented running time tracking for performance analysis.

2. **halos.py**:
   - Added running time metrics for enhanced performance tracking.

3. **Notebook Renaming**:
   - Renamed `get_halos.ipynb` to `halos_rendering.ipynb`.

4. **halos_rendering.ipynb Enhancements**:
   - Added `kappa_gamma_distribution` for visualizing the joint distribution of kappa-gamma_tot.
   - Introduced visualizations to showcase the distributions of the modified kappa and gamma values from halos using histograms.
   - Incorporated KDE functionality for generating random samples of `kappa_halos` and `gamma_halos`.

5. **Data Files**:
   - Added `kgdata.npy`: Weak-lensing maps sourced from GLASS with n=128 (z=5). (better try other way to upload)
   - Introduced `z5data.npy`: Weak-lensing maps from GLASS with n=2048 (z=5).

6. **halos_plus_glass.py**:
   - This script fuses several functions to produce the joint distribution of kappa and gamma values leveraging both halo and GLASS data.
     a. **Glass Data & Sample Generation**: Utilized `read_glass_data` to extract kappa and gamma values from the Glass data file. Implemented `skyarea_form_n` to deduce the corresponding sky area for each pixel. Employed `generate_samples_from_glass` to fit a Gaussian KDE to the joint distribution and generate a random sample.
     b. **Halo Data & Sample Generation**: Applied a Gaussian KDE to derive the joint distribution and presented a hexbin plot visualization.
     c. **Combining Halo and Glass Data**: Invoked `generate_halos_multiple_times` to repeatedly generate kappa and gamma values from halos. Subsequently, merged the halo and Glass data samples with the `halos_plus_glass` function.

7. **glass_convergencen_shear.ipynb**:
   - A dedicated notebook to import data from GLASS and facilitate its visualization.

8. **glass_plus_halos_rendering.ipynb**:
   - This notebook illustrates the process of deriving kappa and gamma values from both halo and GLASS sources. It also demonstrates the method to visualize the combined kappa and gamma values' joint distribution.
----------------------------------------------------------------------
If there are no issues with these codes and the underlying physics, I will generate weak lensing maps for add different values of z and add test function next.

sibirrer commented on sim_pipeline/halos/halos_lens.py:
```diff
@@ -0,0 +1,233 @@
+                                    cosmo=self.cosmo,
+                                    observed_convention_index=[],
+                                    multi_plane=True,
+                                    z_source=5
```
the source redshift is hard-coded here. Perhaps have it as an input argument and make sure the halo rendering and LensModel instance class are consistent with each other

sibirrer commented on sim_pipeline/halos/halos_lens.py:
```diff
@@ -0,0 +1,233 @@
+        kappa, gamma1, gamma2 : float
+            The computed convergence and two components of the shear at the origin.
+        """
+        kappa = self.lens_model.kappa(0.0, 0.0, self.get_halos_lens_kwargs(),
```
you might be able to avoid duplicate calculations for kappa and gamma when you instead evaluate the hessian() and then calculate kappa and gamma values directly from it. Have a look at https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L204
And https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L247 , https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L289

sibirrer reviewed (changes requested):
I made some minor suggestions and I identified an issue with the area sampling of the haloes as a function of radius, that might cause you problems with the LOS computation

trivialTZ committed:
Debug and improve code

1. debug: including debug about the position sample (of disk); halos mass's resolution and others.
2. improve: add/improve some methods and improve the docstring.
3. The results should be more realistic now.

trivialTZ committed:
Short-term solve the SkyPy issue and optimize

1.make a duplicate of Skypy halos branch and make a read.me for credits statement
2. Optimize the project. （mulitprocess & numpy vector operator)
(still some room for improve)

trivialTZ committed:
The missing upload code from last commit

(optimized code)

trivialTZ committed:
Make it past test

trivialTZ committed:
Update abundance_matching.py

For passing the test

trivialTZ committed:
update for passing test

trivialTZ committed:
Update halo.yml for passing the test

trivialTZ committed:
Update halo.yml

trivialTZ committed:
Update test_halos_pipeline.py

sibirrer reviewed (changes requested):
Looks great!
Can you move the jupyter notebooks to the notebook folder?
Can you add test cases for all functions that you added, and not just the HalosSkyPyPypeline?
Can you add for all the files/codes that you copy-pasted from SkyPy a link to the original code with a copy-right notice of SkyPy and a TODO that once it is in the main branch or release version of SkyPy, that this code becomes deprecated?

Thank you very much!

trivialTZ committed:
Test function & debug

1. add test function for Halos
2. rename file name halos ---> Halos
3. move the jupyter notebooks to the notebook folder
4. debug of halos_plus_glass.py ()
5. give proper attribution for the code copy from Skypy (as once it is in the main branch or release version of SkyPy, that this code becomes deprecated)

@sibirrer

trivialTZ committed:
Delete glass_convergence_shear_map.ipynb

trivialTZ committed:
Delete halos_rendering.ipynb

trivialTZ committed:
Delete glass_plus_halos_rendering.ipynb

trivialTZ committed:
Update test_requirements.txt

For tqdm

trivialTZ commented:
1. add test function for Halos
2. rename file name halos ---> Halos
3. move the jupyter notebooks to the notebook folder
4. debug of halos_plus_glass.py ()
5. give proper attribution for the code copy from Skypy (as once it is in the main branch or release version of SkyPy, that this code becomes deprecated)

@sibirrer

sibirrer commented on sim_pipeline/Skypy_halos_duplicate/halos/abundance_matching.py:
```diff
@@ -0,0 +1,102 @@
+]
+
+
+def vale_ostriker(halo_kwargs, subhalo_kwargs, galaxy_kwargs, cosmology):
```
do we need this function from skypy?

sibirrer commented on sim_pipeline/Skypy_halos_duplicate/halos/quenching.py:
```diff
@@ -0,0 +1,117 @@
+    ]
+
+
+def environment_quenched(nh, probability):
```
do we need this function?

sibirrer commented on sim_pipeline/Skypy_halos_duplicate/halos/quenching.py:
```diff
@@ -0,0 +1,117 @@
+    return np.random.uniform(size=nh) < probability
+
+
+def mass_quenched(halo_mass, offset, width):
```
do we need this function?

sibirrer reviewed (commented):
I see, there is quite a lot of code in the SkyPy package that needs to be ported. I just heard back that they try to make a new release that could solve this part hopefully soon(ish).

But good having it in for now such that we can test our functionalities on top of it.

trivialTZ committed:
Improve code and debug

1. add def run_halos_without_kde_by_multiprocessing() as it can help improve halos rendering when n_iterations is large.
2. debug

trivialTZ committed:
Add test for new method & delete unnecessary duplicate from Skypy

1.Add test for new method
2.delete unnecessary duplicate from Skypy
3.move pipeline test file
4.minor change on the markdown of a jupyter notebook

trivialTZ committed:
Add mass sheet correction

Test function and documentation are on the way.

trivialTZ committed:
debug on calculating mass expectation

still need further look of what went wrong.

trivialTZ committed:
Incomplete debug

1.deg2_to_cone_angle debug
2.add a test function

trivialTZ committed:
debug for the mass-sheet correction

1.add test for cone_radius_angle_to_physical_area()
2. will update documentation and complete test functions in the furture.

ajshajib commented on notebooks/halos_rendering.ipynb:
```diff
@@ -0,0 +1,457 @@
+{
```
<p>You don't need this line: "The first line of code is an import statement, it brings in the <code>HalosSkyPyPipeline</code> class from the <code>sim_pipeline.Pipelines.halos_pipeline</code> module." It is self-evident just by looking at the code.</p><p> </p><p>Instead, you can say something like "We first create an instance of the <code>HalosSkyPyPipeline</code> ..." and then explain why these values for <code>m_max</code>,<code>m_min</code>, etc. are chosen.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1316635028'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

ajshajib replied:
<p>A scatter plot showing with $z$ along the x-axis and $\log_{10}(\rm mass)$ along the y-axis will illustrate this table nicely.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1316635033'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

ajshajib replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    kappa_gamma_distribution = halos_lens.get_kappa_gamma_distib(gamma_tot=True, diff=None)
</pre><p>The function should be called as <code>get_kappa_gamma_distribution()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/sibirrer/sim-pipeline/pull/7/discussion/#comment-1316635035'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

sibirrer committed:
Merge branch 'main' into main

sibirrer commented:
@trivialTZ I closed this PR as you can directly make a PR against the origin in the LSST repository
